cli/command/container/opts.go

```diff
@@ -348,10 +349,23 @@ func parse(flags *pflag.FlagSet, copts *containerOptions, serverOS string) (*con
 	// add any bind targets to the list of container volumes
 	for bind := range copts.volumes.GetMap() {
 		parsed, _ := loader.ParseVolume(bind)
+
 		if parsed.Source != "" {
+			toBind := bind
```

Comment:
This code is a bit hairy (not your fault), and could use some cleaning up at least (we're parsing the `bind` into a structured type, but effectively only to do some mild "validation" (however, we're throwing away any error caused by `loader.ParseVolume()`.

One thing we _can_ do already (if I see it correctly), is to use the `parsed.Type` field to only run this code if it's a bind-mount.

Ideally (but that bit might be ok to be done separate), we should

- possibly make have `loader.ParseVolume()` already do the resolving (I guess that won't work for the compose case, as it would have to resolve _relative to_ the compose-file)
- if the above is not an option because of compose-file, we could update the `parsed.Source`, and;
- change `types.ServiceVolumeConfig` (which is returned by `loader.ParseVolume()`) to have a `.String()` (or a separate utility for that), which returns `parsed` in its canonical `string` representation.

Reply:
I added the check for the `parsed.Type`, I didn't really want to touch the loader code since it touches the compose/swarm things and figured it would surely break something else. I do agree we could have something somewhere that can return a canonical string representation, we should also try and think about how we can make volume and mounts options parsing be in the same place.